src/Sarif/Baseline/V2/WhatComparer.cs

```diff
@@ -0,0 +1,192 @@
+                        }
+                    }
+                }
+
+                // Force non-match if there were fingerprints but none of them matched
+                if (correspondingFingerprintCount > 0) { return false; }
+            }
+
+            // Match PartialFingerprints (50% must match)
+            if (left.Result.PartialFingerprints != null && right.Result.PartialFingerprints != null)
+            {
+                int correspondingFingerprintCount = 0;
+                int matchCount = 0;
+
+                foreach (var fingerprint in left.Result.PartialFingerprints)
+                {
+                    if (right.Result.PartialFingerprints.TryGetValue(fingerprint.Key, out string otherFingerprint))
+                    {
+                        correspondingFingerprintCount++;
+
+                        if (string.Equals(fingerprint.Value, otherFingerprint))
+                        {
+                            matchCount++;
+                        }
+                    }
+                }
+
+                if (correspondingFingerprintCount > 0)
+                {
+                    // If fingerprints were found, return true when at least half of them matched, false otherwise
+                    return (matchCount > 0 && matchCount * 2 >= left.Result.PartialFingerprints.Count);
+                }
+            }
+
+            // Match Properties (50% must match)
+            if (left.Result.Properties != null && right.Result.Properties != null)
+            {
+                int correspondingPropertyCount = 0;
+                int matchCount = 0;
+
+                foreach (var property in left.Result.Properties)
+                {
+                    if (left.Result.TryGetProperty(property.Key, out string leftValue)
+                        && right.Result.TryGetProperty(property.Key, out string otherPropertyValue))
+                    {
+                        correspondingPropertyCount++;
+
+                        if (string.Equals(leftValue, otherPropertyValue))
+                        {
+                            matchCount++;
+                        }
+                    }
+                }
+
+                if (correspondingPropertyCount > 0)
+                {
+                    if (matchCount == 0 || matchCount * 2 < left.Result.Properties.Count)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            // At this point, no high confidence properties matched or failed to match
+            string leftMessage = left.Result.GetMessageText(left.Result.GetRule(left.OriginalRun));
+            string rightMessage = right.Result.GetMessageText(right.Result.GetRule(right.OriginalRun));
+
+            string leftSnippet = GetFirstSnippet(left);
+            string rightSnippet = GetFirstSnippet(right);
+
+            return string.Equals(leftMessage, rightMessage) && string.Equals(leftSnippet, rightSnippet);
```

Comment:
>string.Equals(leftMessage, rightMessage) && string.Equals(leftSnippet, rightSnippet); [](start = 19, length = 85)

I wouldn't have expected the message and snippets to be in the what comparison.

Reply:
Yeah, some Result properties could arguably belong to "what" or "where". I wanted to restrict where to the location and not the content, so that very simple renames or code changes would still result in a where match. The Message may be static for the rule, but will sometimes describe what is wrong (this class didn't properly implement interface ISomethingOrOther), in which case it would describe what the problem is. The Snippet hopefully includes all of the code encompassing the problem, when it's not a "this is missing" problem, so it also can potentially describe what the problem is.

I do think the "what" and "where" sets and the strictness of matching we require is something we'll likely iterate on. For example, if the line numbers stay but the column numbers move (format document indenting), should we still consider two Results a where match? Probably.

Reply:
I agree with Scott that snippet and message are more appropriately "what" than "where".

---
In reply to: [295894088](https://github.com/microsoft/sarif-sdk/pull/1539#discussion_r295894088) [](ancestors = 295894088)